### DISCUSSION TITLE
Use `use_displaced_mesh` in `StressDivergenceTensors`

### DISCUSSION THREAD

**Discussion by HemiaoZhang** [idaholab/moose#24804] [Q&A Modules: Solid mechanics] [answered]
Hi moose community,

I have three questions regarding the kernel calculation based on current/reference in `StressDivergenceTensors` for finite strain plasticity.

The first one is whether the gradient of the test function is taken w.r.t. the current or reference configuration? I tried to trace the definition for ` _grad_test` to `Kernel.C` where it is defined in the initialization list as,

```c++
_test(_var.phi()),
_grad_test(_var.gradPhi()),
```

However, I was unable to find detailed documentation for `gradPhi()` except for it being a member function of `MooseVariableData`. Therefore, could you suggest which `gradPhi()` is taken w.r.t to?

The second question is about the `use_displaced_mesh`. From the [note](https://mooseframework.inl.gov/moose/source/kernels/StressDivergenceTensors.html), I suppose it is a switch to control which configuration to use. However, from the code snippet below, it seems that such switch is only applicable when the mesh dimension is less than 3. 

```c++
Real
StressDivergenceTensors::computeQpResidual()
{
  Real residual = _stress[_qp].row(_component) * _grad_test[_i][_qp];
  // volumetric locking correction
  if (_volumetric_locking_correction)
    residual += _stress[_qp].trace() / 3.0 *
                (_avg_grad_test[_i][_component] - _grad_test[_i][_qp](_component));

  if (_ndisp != 3 && _out_of_plane_strain_coupled && _use_displaced_mesh)
  {
    const Real out_of_plane_thickness = std::exp((*_out_of_plane_strain)[_qp]);
    residual *= out_of_plane_thickness;
  }

  return residual;
}
```

Furthermore, adding `use_displaced_mesh = true` to the master action results in an error.

```
*** ERROR ***
/Users/hemiaozhang/projects/finite_plasticity/problems/finite_dp_ver1.i:30: unused parameter 'Modules/TensorMechanics/Master/all/use_displaced_mesh'
```

My last question is, therefore, which stress should we feed the `StressDivergenceTensors`? In other words, when we are defining a new material, which stress should be the final result?

Many thanks in advance.

**Comment by GiudGiud** [comment]
@recuero @hugary1995

**Comment by hugary1995** [comment, marked as answer]
> The first one is whether the gradient of the test function is taken w.r.t. the current or reference configuration?

The gradient is wrt the current configuration if `use_displaced_mesh = true`, wrt reference configuration otherwise.

> The second question ... However, from the code snippet below, it seems that such switch is only applicable when the mesh dimension is less than 3.

That's not true.

> Furthermore, adding use_displaced_mesh = true to the master action results in an error.

The action will set the appropriate configuration for you based on other parameters. Hence the error.

> My last question is, therefore, which stress should we feed the StressDivergenceTensors? In other words, when we are defining a new material, which stress should be the final result?

The Cauchy stress.

**Reply by HemiaoZhang** [reply]
Thank you so much for the answers!

I have a further question. Could you help suggest which setting in the input file might conflict with `use_displaced_mesh = true`? The input file is listed pasted.

```
[Mesh]
  [generated_mesh]
    type = GeneratedMeshGenerator
    elem_type = HEX8
    dim = 3
    nx = 1
    ny = 1
    nz = 1
    xmin = 0.0
    xmax = 1.0
    ymin = 0.0
    ymax = 1.0
    zmin = 0.0
    zmax = 1.0
  []
[]

[GlobalParams]
  displacements = 'disp_x disp_y disp_z'
[]

[Modules/TensorMechanics/Master]
  [all]
    strain = FINITE
    displacements = 'disp_x disp_y disp_z'
    decomposition_method = EigenSolution
    #use_displaced_mesh = true
    add_variables = true
    use_finite_deform_jacobian = true
    volumetric_locking_correction = false
    generate_output = 'stress_xx stress_yy stress_zz stress_xy stress_xz'
  []
[]

[Materials]
  [stress]
    type = ComputeFiniteDruckerPragerStress
    displacements = 'disp_x disp_y disp_z'
    cohesion_c = 1e+7        # [Pa]
    friction_phi = 0.5236    # [rad]
    dilatancy_psi = 0        # [rad]
    circle_choice = Inner
    hardening_n = 1e+15
    lame_lambda = 8.0769e+9  # [Pa]
    lame_mu = 5.3846e+9      # [Pa]
    elasticity_model = Hencky
    etol = 1e-10
    ptol = 1e-10
    ftol = 1e-10
  []
  [elasticity_tensor]
    type = ComputeIsotropicElasticityTensor
    lambda = 8.0769e+9        # [Pa]
    shear_modulus = 5.3846e+9 # [Pa]
  []
[]

[BCs]
  [top]
    type = FunctionDirichletBC
    variable = disp_z
    boundary = top
    function = '0.001 * t'
  []
  [bottom_x]
    type = DirichletBC
    variable = disp_x
    boundary = bottom
    value = 0.0
  []
  [bottom_y]
    type = DirichletBC
    variable = disp_y
    boundary = bottom
    value = 0.0
  []
  [bottom_z]
    type = DirichletBC
    variable = disp_z
    boundary = bottom
    value = 0.0
  []
[]

[Preconditioning]
  [./SMP]
   type = SMP
   full=true
  [../]
[]

[Executioner]
  type = Transient
  dt=1
  end_time=10
  #nl_abs_tol = 1e-8
[]

[Outputs]
  [exodus]
    type = Exodus
  []
[]

[Debug]
  show_material_props = true
[]
```

Also, since the algorithm I have is based on the total strain not the incremental strain, I added `displacements = 'disp_x disp_y disp_z'` in `[Materials/stress]`, will it possibly be inconsistent with `ComputeFiniteStrain`?

And one last general question on the `Executioner`. In the output, there is usually a linear solver nested within a nonlinear one. I can understand that the nonlinear solver is to solve for displacement from the global stiffness matrix and the right hand side. What is the linear solver for? I saw from previous discussion that the convergence issue is usually with the nonlinear solver and not the linear one and am having the same issue myself. What could possibly cause that?

**Reply by hugary1995** [reply]
> I have a further question. Could you help suggest which setting in the input file might conflict with use_displaced_mesh = true? The input file is listed pasted.

What do you mean by "conflict with use_displaced_mesh = true"? Where and why do you want to manipulate use_displaced_mesh yourself. We typically advise against that.

> Also, since the algorithm I have is based on the total strain not the incremental strain, I added displacements = 'disp_x disp_y disp_z' in [Materials/stress], will it possibly be inconsistent with ComputeFiniteStrain

This question along with the previous one are very difficult to answer in general. Please clarify:
1. Are you dealing with small deformation or large deformation?
2. If large deformation, which strain/stress measures are you working with?
3. Why do you insist on computing the stress on the deformed configuration?

> And one last general question on the Executioner. In the output, there is usually a linear solver nested within a nonlinear one. I can understand that the nonlinear solver is to solve for displacement from the global stiffness matrix and the right hand side. What is the linear solver for?

No, the linear system is the traditional K u = f taught in old-school engineer classes. For a general nonlinear system, R(u) = 0, you need to iteratively update it (hence the nonlinear loop). See https://en.wikipedia.org/wiki/Newton%27s_method

> I saw from previous discussion that the convergence issue is usually with the nonlinear solver and not the linear one and am having the same issue myself. What could possibly cause that?

This is again very difficult to answer because the cause of convergence issues can vary dramatically from problem to problem. The troubleshooting page should be a good starting point: https://mooseframework.inl.gov/application_usage/failed_solves.html
